File: 2022/dec16.js
const readline = require('readline');

let rl = readline.createInterface({
  input: process.stdin
});

let nodes = [];
let names = [];
let done = {};
let numCalls = 0;
rl.on('line', function (line) {
  let name = line.substring(6, line.length).split(" has")[0];
  line = line.split("rate=")[1];
  let rate = parseInt(line.split(":")[0]);
  line = line.split("valve")[1];
  let edges = [];
  if (line[0] === "s") line = line.substring(1, line.length);
  line.split(",").forEach(node => {
    edges.push(node.substring(1, node.length));
  });
  names.push(name);
  nodes[name] ={
    "rate": rate,
    "edges": edges
  };
});

const getDists = (node, list, nodes) => {
  let ret = {};
  let visited = [node];
  let q = [[node,0]]
  while(q.length > 0) {
    let curr = q.shift();
    nodes[curr[0]].edges.forEach(n => {
      if (!visited.includes(n)) {
        q.push([n, curr[1] + 1]);
        visited.push(n);
        if (list.includes(n)) ret[n] = curr[1] + 1;
      }
    });
  }
  return ret;
}

const getMax = (pos, dists, times, val, unopen) => {
  if (Math.max(times) < 1) {
    return val;
  }

  let idx = times.indexOf(Math.max(...times));
  let k1 = pos + times + unopen;
  let k2 = [...pos].reverse() + [...times].reverse() + unopen;
  if (done[k1]) {
    return Math.max(done[k1], val);
  } else if (done[k2]) {
    return Math.max(done[k2], val);
  }

  let curr = [[val, '']];
  if (unopen.includes(pos[idx])) {
    let newT = [...times];
    let newPos = [...pos];
    let newUnopen = [...unopen].filter(n => n !== pos[idx]);
    newT[idx] = newT[idx] - 1;
    let newVal = val + newT[idx] * nodes[pos[idx]].rate;
    curr.push([getMax(newPos, dists, newT, newVal, newUnopen), pos[idx]]);
  }

  for (const [key, value] of Object.entries(dists[pos[idx]])) {
    if (unopen.includes(key) && times[idx] > value + 1) {
      let newT = [...times];
      let newPos = [...pos];
      let newUnopen = [...unopen].filter(n => n !== key);
      newT[idx] = newT[idx] - (value + 1);
      newPos[idx] = key;
      let newVal = val + newT[idx] * nodes[key].rate;
      curr.push([getMax(newPos, dists, newT, newVal, newUnopen), key]);
    }
  };
  let m = Math.max(...curr.map(c => c[0]));
  if (unopen.length > 1) {
    done[k1] = m;
  }
  return m;
}

rl.on('close', function (cmd) {
  let start = "AA";
  let interesting = [];

  names.forEach(name => {
    if (nodes[name].rate > 0) interesting.push(name);
  });

  // Find shortest distance for each node with rate > 0 to each other node with rate > 9
  let dists = {};
  [start, ...interesting].forEach(name => {
    dists[name] = getDists(name, [start, ...interesting], nodes);
  });

  let ans1 = getMax([start], dists, [30], 0, [start, ...interesting], {});
  console.log("Part 1 - " + ans1);
  console.log("JS can't handle the needed recursion depth for problem 2");
});

File: 2022/dec16.py
import re
import sys
import math
import time
from copy import deepcopy
from collections import deque

def getDists(node, interesting, valves):
    ret = {}
    v = [node]
    q = [(node, 0)]
    while len(q) > 0:
        curr = q.pop(0)
        for next in valves[curr[0]]['conns']:
            if not next in v:
                v.append(next)
                q.append((next, curr[1] + 1))
                if next in interesting:
                    ret[interesting.index(next)] = curr[1] + 1
    return ret

valves = {}
lines = open(sys.argv[1]).read().split('\n')
for line in lines:
    p1, p2 = line.split('; ')
    arr1 = p1.split(' ')
    name = arr1[1]
    flow = int(re.findall(r'\d+', arr1[4])[0])
    arr2 = p2.split(' ')[4:]
    conns = [s[0:2] for s in arr2]
    valves[name] = {'flow': flow, 'conns': conns}

dists = []
flows = []
interesting = ['AA'] + [k for k, x in valves.items() if x['flow'] > 0]
for i, node in enumerate(interesting):
    dists.append(getDists(node, interesting, valves))
    flows.append(valves[node]['flow'])

seen = {}
def findMax(pos, t, unopen):
    findMax.counter += 1
    if findMax.counter % 1000000 == 0:
        print(findMax.counter)

    if max(t) < 3:
        return 0

    if (*pos, *t, unopen) in seen:
        return seen[(*pos, *t, unopen)]
    if (*pos[::-1], *t[::-1], unopen) in seen:
        return seen[(*pos[::-1], *t[::-1], unopen)]

    idx = t.index(max(t))
    ps = [0]
    for next, dist in dists[pos[idx]].items():
        bit = 1 << next
        if not unopen & bit and t[idx] > dist + 2:
            newT = [*t]
            newPos = [*pos]
            newT[idx] = t[idx] - (dist + 1)
            newPos[idx] = next
            ps.append(newT[idx] * flows[next] + findMax(newPos, newT, unopen | bit))

    seen[(*pos, *t, unopen)] = max(ps)
    return max(ps)

findMax.counter = 0
ans1 = findMax([0], [30], 1) # 1 means 'AA' is open since flow = 0
print(f'Answer Part 1: {ans1}')

seen = {}
start = time.time()
ans2 = findMax([0, 0], [26, 26], 1)
end = time.time()
print(f'Answer Part 2: {ans2} it took: {end-start}')